### PR TITLE
feat(model): add Gemma-4 E4B support (layer spec, checkpoint loader, parity check)

### PR DESCRIPTION
**Pull request by DOGEUNNKIM**
# What does this PR do?

- Adds full Gemma-4 E4B (3.8B dense text) support to Megatron-Bridge,
  consolidating all Gemma4-specific code that was previously scattered in
  Megatron-LM core (see MCore PR: https://github.com/NVIDIA/Megatron-LM/pull/5090).
- Introduces `Gemma4DenseProvider` — a Bridge-native model provider that works
  against **clean Megatron-Core** (no Gemma4-specific CLI args or
  `TransformerConfig` fields required in MCore).
- Separates the text-only path (`Gemma4Bridge` / `Gemma4ForCausalLM`) from the
  VL path (`Gemma4VLBridge` / `Gemma4ForConditionalGeneration`); `gemma_vl`
  imports from `gemma`, not vice versa.
- Provides HF → Megatron checkpoint conversion, logit parity verification, and
  training launch scripts.

# Changelog

#### `src/megatron/bridge/models/gemma/modeling_gemma4.py`
Core Gemma-4 architecture:
- **`Gemma4TransformerLayer`** — 4-norm residual structure, Phase 4 PLE residual
  block (`gelu(gate(h)) × per_layer_input → proj → norm → residual`), per-layer
  scalar, optional MoE block.
- **`Gemma4SelfAttention`** — GQA with per-head `q_norm`/`k_norm`,
  heterogeneous head_dim (sliding=256, global=512), shared-KV forwarding.
- **`Gemma4DenseRotaryEmbedding`** — Dual-theta RoPE: sliding layers use
  θ=10 000, global layers use θ=1 000 000 with partial rotation (factor=0.25).
- **`wire_gemma4_kv_sharing()`** — Post-construction wiring of shared-KV source
  references for the last `num_kv_shared_layers` layers.

#### `src/megatron/bridge/models/gemma/gemma4_provider.py`
- **`Gemma4DenseProvider`** — All-in-one dataclass provider for clean MCore:
  builds `GPTModel`, attaches dual RoPE, PLE modules, and wires KV sharing.
  `activation_func` uses `fast_gelu` (registered as `"gelu_pytorch_tanh"` in
  `ACTIVATION_FUNC_MAP`) so `megatron_to_hf_activation` can match it by
  identity — avoids `"Unsupported activation function"` on HF export.

#### `src/megatron/bridge/models/gemma_vl/modeling_gemma4_vl.py`
- **`Gemma4VLModel`** — Wraps HF vision/audio towers with Megatron language
  model.
  - `_compute_attention_mask` returns a `(B, 1, S, S)` bool tensor
    (`True = blocked`) instead of a dict, matching unit-test expectations and
    downstream pipeline usage.
  - `forward()` returns logits directly in inference mode
    (`labels=None, loss_mask=None`) to avoid `tuple` `AttributeError` in the
    pipeline forward pass.

#### `src/megatron/bridge/models/gemma/gemma4_bridge.py` / `src/megatron/bridge/models/gemma_vl/gemma4_vl_bridge.py`
- `Gemma4Bridge` handles `Gemma4ForCausalLM`; includes a load-state alias hook
  that maps `self_attention_sliding/global.*` checkpoint keys to
  `self_attention.*` model keys, silencing the strict-load warning on export.
- `Gemma4VLBridge` handles `Gemma4ForConditionalGeneration`; respects
  `GEMMA4_CONVERSION_MODE=text|vl|audio` to select the appropriate provider
  and parameter mappings.

#### `examples/models/gemma/gemma4/conversion.sh`
HF → Megatron import, Megatron → HF export, and round-trip parity check.
Sets `GEMMA4_CONVERSION_MODE=text` to use `Gemma4DenseProvider` (GPTModel)
for `gemma-4-E4B-it` which is otherwise detected as
`Gemma4ForConditionalGeneration`.

#### `examples/models/gemma/gemma4/inference.sh`
Text generation from HF or Megatron checkpoint using
`hf_to_megatron_generate_text.py`. Sets `GEMMA4_CONVERSION_MODE=text`; prompts
use the Gemma 4 IT chat template for the instruction-tuned model.

#### `examples/models/gemma/gemma4/parity_check_e4b.py`
Distributed logit parity check (TP=1/2). Expected max|diff| < 3.0 (bf16).

#### `examples/models/gemma/gemma4/README.md`
Usage guide: requirements, workspace setup, conversion, inference, parity
checks, pretraining, and architecture notes.

## Relation to MCore PR

This PR is the **Bridge half** of the split requested in
[Megatron-LM #5090](https://github.com/NVIDIA/Megatron-LM/pull/5090).
The only MCore dependency retained is the generic `per_layer_inputs` hook
already threaded through `TransformerBlock.forward()`.

# Validation

Validated locally with:

- HF → Megatron import ✅
- Megatron → HF export ✅
- Round-trip weight verification ✅
- Text,vision,audio modality logit parity check ✅
- Unit tests: all 227 tests passing ✅

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
- [ ] Reviewer: Does the PR have correct import guards for all optional libraries?

### CODE REVIEW

File: src/megatron/bridge/models/gemma/gemma4_bridge.py [hunk 1]
```diff
@@ -1,541 +0,0 @@
...
-    skipping the entire QKV mapping; the V weights are synthesized from K
-    in ``Gemma4Bridge.maybe_modify_loaded_hf_weight``.
-    """
-
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
-        self.allow_hf_name_mismatch = True
-
-
-@MegatronModelBridge.register_bridge(
```
Comment (yaoyu-33) [inline src/megatron/bridge/models/gemma/gemma4_bridge.py:88]: probably do not remove these? This register to pure text model, still useful in many cases.

File: examples/models/gemma/gemma4/parity_check_e4b.py [hunk 2]
```diff
@@ -0,0 +1,543 @@
+#!/usr/bin/env python3
```
Comment (yaoyu-33) [inline examples/models/gemma/gemma4/parity_check_e4b.py:1]: this is good. on bridge side, it's better if we can also do a  conversion.sh and a inference.sh - that shows the conversion works and generation results are coherent.

File: src/megatron/bridge/models/gemma_vl/gemma4_vl_bridge.py [hunk 3]
```diff
@@ -46,145 +53,183 @@
...
+# Gemma4Bridge — text-only CausalLM bridge (MoE and Dense)
+# ---------------------------------------------------------------------------
+
+
 @MegatronModelBridge.register_bridge(
-    source="Gemma4ForConditionalGeneration",
-    target=Gemma4VLModel,
-    provider=Gemma4VLModelProvider,
-    model_type="gemma4_vl",
+    source="Gemma4ForCausalLM",
```
Comment (yaoyu-33) [inline src/megatron/bridge/models/gemma_vl/gemma4_vl_bridge.py]: oh pure llm should not be merged into here, you can levarage the class method if needed.

File: src/megatron/bridge/models/gemma/gemma4_bridge.py (hunk 1 above)
Comment (DOGEUNNKIM) [reply src/megatron/bridge/models/gemma/gemma4_bridge.py:88]: Agreed — kept as-is. Gemma4ForCausalLM registration in Gemma4Bridge is preserved for text-only use cases.

File: examples/models/gemma/gemma4/parity_check_e4b.py (hunk 2 above)
Comment (DOGEUNNKIM) [reply examples/models/gemma/gemma4/parity_check_e4b.py:1]: Done — added examples/models/gemma/gemma4/conversion.sh (HF → Megatron import, Megatron → HF export, round-trip parity check) and inference.sh (text generation with chat template).

File: src/megatron/bridge/models/gemma_vl/gemma4_vl_bridge.py (hunk 3 above)
Comment (DOGEUNNKIM) [reply src/megatron/bridge/models/gemma_vl/gemma4_vl_bridge.py]: Agreed — pure LLM code lives entirely under models/gemma. models/gemma_vl only contains multimodal wrappers (Gemma4VLModel), VL providers (Gemma4DenseVLProvider, Gemma4VLModelProvider), and VL bridge mappings.

**Comment by DOGEUNNKIM** [comment]
@yaoyu-33  Thanks for the review. I updated the PR based on the feedback.

- Kept `Gemma4ForCausalLM` registration in `Gemma4Bridge` for pure text use cases.
- Added `examples/models/gemma/gemma4/conversion.sh` and `inference.sh`.
- Refactored pure LLM code out of `models/gemma_vl`.

The shared/text Gemma4 implementation now lives under `models/gemma`:
- `modeling_gemma4.py`
- `gemma4_provider.py`
- `gemma4_bridge.py`

`models/gemma_vl` now only keeps multimodal-specific wrappers, VL providers, and VL/audio mappings, while reusing the text provider/modeling code from `models/gemma`.

I also fixed the Gemma4 Dense export-load warning caused by `self_attention_sliding/global.*` checkpoint keys versus `self_attention.*` model keys using a Gemma4-only load-state alias hook.

Local validation:
- HF -> Megatron import: passed
- Megatron -> HF export: passed
- round-trip weight verification: passed
- text ,vision, audio logit parity: passed

File: src/megatron/bridge/recipes/gemma/gemma4.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    Override at launch time with Hydra-style args, e.g.::
+
+        checkpoint.pretrained_checkpoint=/path/to/megatron-ckpt
+        checkpoint.save=/path/to/save
+        train.train_iters=1000
+        model.seq_length=4096
+    """
+    cfg = _pretrain_common()
+
+    cfg.model = Gemma4DenseProvider(
```
Comment (yaoyu-33) [inline src/megatron/bridge/recipes/gemma/gemma4.py]: This recipe manually duplicates the Gemma4 E4B architecture instead of deriving it through the bridge path. That can be valid for a fixed pretrain recipe, but this needs a focused recipe test that asserts the critical provider fields and catches drift from `Gemma4Bridge._build_dense_provider()`.

File: examples/models/gemma/gemma4/slurm_pretrain.sh [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+}
+
+# ---------------------------------------------------------------------------
+# Helper: run one parity check
+# ---------------------------------------------------------------------------
+_parity() {
+    local mode="$1"
+    local ckpt_path="$2"
+    local port="$3"
+    local log_dir="${GEMMA4_LOG_ROOT:-/mnt/nvme0/kdg6245}/gemma4_e4b_parity_${mode}"
```
Comment (yaoyu-33) [inline examples/models/gemma/gemma4/slurm_pretrain.sh]: Please remove the `/mnt/nvme0/kdg6245` default. Public examples should not carry user/environment-specific paths; default to a repo-neutral results/log directory or require `GEMMA4_LOG_ROOT`.

File: examples/models/gemma/gemma4/slurm_pretrain.sh [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+# ---------------------------------------------------------------------------
+# Validation
+# ---------------------------------------------------------------------------
+if [ ! -d "$HF_MODEL_DIR" ]; then
+    echo "Error: HF model not found at $HF_MODEL_DIR"
+    echo "  Download with: huggingface-cli download google/gemma-4-E4B-it --local-dir $HF_MODEL_DIR"
+    exit 1
+fi
+
+TORCHRUN_BIN=${TORCHRUN_BIN:-torchrun}
```
Comment (yaoyu-33) [inline examples/models/gemma/gemma4/slurm_pretrain.sh]: Repo guidance uses `uv run python -m torch.distributed.run` rather than bare `torchrun`. Please update this default launcher and the matching README commands so users run in the uv-managed environment.

File: examples/models/gemma/gemma4/parity_check_e4b.py [hunk 7]
```diff
@@ -0,0 +1,583 @@
...
```
Comment (yaoyu-33) [inline examples/models/gemma/gemma4/parity_check_e4b.py:1]: This new non-test Python file needs the standard 2026 NVIDIA copyright header.

**Review by yaoyu-33 — COMMENTED** [review]
Structured review comments from local review.

File: examples/models/gemma/gemma4/parity_check_e4b.py (hunk 7 above)
Comment (DOGEUNNKIM) [reply examples/models/gemma/gemma4/parity_check_e4b.py:1]: Done - Added NVIDIA 2026 Apache 2.0 copyright headers to all new non-test Python files.

File: src/megatron/bridge/recipes/gemma/gemma4.py (hunk 4 above)
Comment (DOGEUNNKIM) [reply src/megatron/bridge/recipes/gemma/gemma4.py]: Agreed - Added tests/unit_tests/recipes/test_gemma4_recipe.py that checks 23 critical model config fields stay in sync between the recipe and the bridge — so any silent drift gets caught automatically.

File: examples/models/gemma/gemma4/slurm_pretrain.sh (hunk 6 above)
Comment (DOGEUNNKIM) [reply examples/models/gemma/gemma4/slurm_pretrain.sh]: Done - Updated TORCHRUN_BIN default to uv run python -m torch.distributed.run to match the repo-wide convention.

File: examples/models/gemma/gemma4/slurm_pretrain.sh (hunk 5 above)
Comment (DOGEUNNKIM) [reply examples/models/gemma/gemma4/slurm_pretrain.sh]: Done - Replaced hardcoded /mnt/nvme0/kdg6245 paths with ${GEMMA4_LOG_ROOT:?'Error: set GEMMA4_LOG_ROOT to ...'} guards so the script fails with a clear message instead of silently using a wrong path.

**Comment by DOGEUNNKIM** [comment]
@yaoyu-33 Thanks for the thorough review! I've addressed all the comments in [cc947371]:

#### Copyright headers: 
Added NVIDIA 2026 Apache 2.0 headers to all new non-test Python files (gemma4_bridge.py, gemma4_provider.py, modeling_gemma4_vl.py, gemma4_vl_bridge.py, gemma4.py, parity_check_e4b.py).

#### Hardcoded paths:
Replaced /mnt/nvme0/kdg6245 defaults in slurm_pretrain.sh with ${VAR:?'Error: set VAR to a writable directory'} guards — the script now fails explicitly if the variable is unset rather than silently using a personal path.

####torchrun → uv: 
Updated TORCHRUN_BIN default to uv run python -m torch.distributed.run per repo convention.

#### Recipe unit test: 
Added tests/unit_tests/recipes/test_gemma4_recipe.py that verifies 23 critical model config fields stay in sync between gemma4_e4b_pretrain_config() and Gemma4Bridge._build_dense_provider(), catching silent drift at test time.

**Comment by DOGEUNNKIM** [comment]
@yaoyu-33 

Updated this PR following the direction from NVIDIA/Megatron-LM#5090 at [9d581da].

The Gemma4 `per_layer_inputs` / PLE threading is now implemented as a Bridge-side compatibility patch instead of requiring Gemma4-specific MCore hooks. 
The patch is scoped to the Gemma4 decoder instance only, so clean MCore behavior remains unchanged.

Please see NVIDIA/Megatron-LM#5090 for the related MCore discussion.

File: src/megatron/bridge/recipes/gemma/gemma4.py (hunk 4 above)
Comment (yaoyu-33) [inline src/megatron/bridge/recipes/gemma/gemma4.py]: Can we just use autobirdge api here as other model recipes? unless there is a strong reason of calling provider directly?

**Comment by yaoyu-33** [comment]
@DOGEUNNKIM  lgtm, had small comment on recipe

File: src/megatron/bridge/recipes/gemma/gemma4.py (hunk 4 above)
Comment (DOGEUNNKIM) [reply src/megatron/bridge/recipes/gemma/gemma4.py]: @yaoyu-33 I updated the Gemma4 E4B pretrain recipe to use `AutoBridge.from_hf_pretrained(...).to_megatron_provider(load_weights=False)` instead of instantiating `Gemma4DenseProvider` directly.

Since `google/gemma-4-E4B-it` goes through the HF `ConditionalGeneration` path, I also scoped `GEMMA4_CONVERSION_MODE=text` during provider creation so the recipe still selects the text-only `Gemma4DenseProvider`.

I updated the recipe unit test accordingly.

**Comment by DOGEUNNKIM** [comment]
> @DOGEUNNKIM lgtm, had small comment on recipe

@yaoyu-33 Thanks, addressed the recipe comment and updated the unit test accordingly.

**Review by yaoyu-33 — DISMISSED** [review]

**Comment by DOGEUNNKIM** [comment]
@yaoyu-33 

Hi, this looks like a flaky docs linkcheck failure rather than an issue caused by this PR.

Sphinx linkcheck marked `https://www.conventionalcommits.org/en/v1.0.0/` as broken due to `Network is unreachable`, but the URL itself appears to be valid. Could we rerun the failed docs job?

Thank you.

**Comment by gautham-kollu** [comment]
@DOGEUNNKIM  gentle reminder of the merge conflicts

**Comment by DOGEUNNKIM** [comment]
> @DOGEUNNKIM gentle reminder of the merge conflicts

@gautham-kollu Thanks for the reminder. I resolved the merge conflicts and pushed the updated changes.

**Comment by ko3n1g** [comment]
/ok to test [2715f5d](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/4148/commits/2715f5d3e010c5e91de8823f23c833d6b1320e1e)

**Comment by DOGEUNNKIM** [comment]
@weijiac0619 

Hi, all CI checks are passing now.

Could you please review this again when you have a chance?

Thank you.

**Comment by weijiac0619** [comment]
sure. taking a look now

**Comment by weijiac0619** [comment]
Hi @DOGEUNNKIM , LGTM. Could you resolve the conflicts?

**Review by weijiac0619 — DISMISSED** [review]

**Comment by DOGEUNNKIM** [comment]
> Hi @DOGEUNNKIM , LGTM. Could you resolve the conflicts?

@weijiac0619 
The conflicts have been resolved by @yaoyu-33. 
Could you please rerun the tests when you have a chance?

Thank you!

**Comment by yaoyu-33** [comment]
previous ci passed, only update for conflicts. force merged.
